### ISSUE TITLE
Scaling by shader

### ISSUE THREAD

**Issue by cactusanctuary** [openfl/openfl#1128]
Hi, I'm putting in a request for a shader which scales the final pixel output of the application. This has been implemented in HaxeFlixel: [pull request](https://github.com/HaxeFlixel/flixel/pull/1826) and [original issue](https://github.com/HaxeFlixel/flixel/issues/1817). 

The goal is to achieve smoothly blended pixels when displaying a small app resolution to a large display resolution. The problem with scaling up everything using `scaleX` and `scaleY` is that it leads to non-square pixels. Applying a shader as a post-processing effect should achieve square pixels by blending the "in-between" pixels.

**Example**
My game's natural resolution is 384x240 pixels. I want to render it to 1080p, while preserving square pixels, so the game should be scaled up by 4.5, which would result in a final resolution of 1728x1080. I can then center and pillarbox the result myself. 

**Targets**
I need it for native targets, specifically Windows and Mac.
Eventually Android would be nice too. 

If anything is unclear, or need more specific info, please ask! :)

**Comment by MSGhero** [comment]
I'm a bit confused. Where would this belong in openfl? It seems more like something you implement yourself or find in another lib.

**Comment by cactusanctuary** [comment]
I don't know where it would belong. I had asked @larsiusprime about it, and he thought it would be a good feature to have available to all OpenFL. 

I don't know of an appropriate place to place it within the Flash API. So maybe it would be better to have it in a separate lib.

**Comment by MSGhero** [comment]
A lot of the things we do in flixel would be really good in like... "openfl-useful" haxelib. For example, I don't think any of the gamepad IDs belong in flixel since those are library-independent. Yet I think that's the only documented place where you can find them all.

**Comment by larsiusprime** [comment]
I have to disagree here MSGhero, I think being able to say "Here are the pixels I have on screen, I would like to scale that up, please" is a very common use case and seems like something that should exist at the OpenFL, or perhaps lime level. Having to implement it again and again for each framework / individual project seems like a waste.

**Comment by cactusanctuary** [comment]
@larsiusprime any ideas of where to place it in openfl API? Or maybe even Lime API?

**Comment by larsiusprime** [comment]
lime.ui.window or openfl.display.window seem like the most natural places to me.

**Comment by cactusanctuary** [comment]
This looks promising: http://community.openfl.org/t/help-opengl-shaders-bitmapdata-scaling-ibilon-base-code-solved/1143/7

The author yupswing shared his scaler shader code and said to try it out. I'm not sure when I can get to it, but hopefully sometime within the week. 

The decision regarding whether it's in Lime or OpenFL, should that be taken by Joshua maybe?

**Comment by larsiusprime** [comment]
Yeah, architecture decisions are mostly up to @jgranick

**Comment by cactusanctuary** [comment]
Actually yupswing's implementation has OpenFL dependencies so it would have to be in there

**Comment by jgranick** [comment]
I'm trying to understand how this works, and if it was core, where it would fit.

If you `scaleX`/`scaleY`, the issue is that it is smoothed, and you want it to be pixelated? Disregarding performance for a moment, would `bitmapData.draw`, with smoothing set to false, then scaling, achieve the correct visual effect?

Would this (possibly) work with `cacheAsBitmapMatrix`? Or is there the same issue of smoothing?

**Comment by cactusanctuary** [comment]
I **want** the smoothing, so that I can fill as much as possible of a screen which isn't an int scale factor. But I get non-square "pixels" because there's no smoothing. 

@larsiusprime's point is that such scaling with smoothness should be part of the OpenFL or Lime core, because it's a pretty common thing to want done. And it's silly for different developers to keep re-implementing it in each of their projects. 

I haven't tried any of those things (`bitmapData.draw` with `smoothing=false`, or `cacheAsBitmapMatrix`) but should I even try them, if you thought I wanted pixelated instead of smoothed?

**Comment by jgranick** [comment]
Okay, so if you want smoothing, I'm confused, because I thought that `scaleX`/`scaleY` used smoothing by default, such as scaling a `Bitmap` with `smoothing = true`. Are you scaling something else that does not have smoothing?

**Comment by MSGhero** [comment]
The flixel issue made it so you could define your shader scale mode. The two options in the PR are nearest neighbor and bilinear. Maybe `smoothing = true` overlaps with NN, but I figure other algos could be added like anything here https://en.wikipedia.org/wiki/Image_scaling#Pixel_art_scaling_algorithms

**Comment by jgranick** [comment]
If I don't misunderstand the issue, currently, `bitmap.smoothing = true/false` switches between bilinear and nearest-neighbor scaling, though other methods could be interesting through custom shaders

**Comment by jgranick** [comment]
Perhaps the issue is that the code is probably going through `drawTiles`, which may not provide this control. Could we do some tests using `Bitmap`/`BitmapData` and see if `smoothing` plus `scaleX`/`scaleY` gives us the control we want?

If it does, then we could then look at how we apply the same functionality to batched quad rendering

Thanks :smile:

**Comment by cactusanctuary** [comment]
Actually I'm using `tilelayer` lib for most sprites in my game, I'm not sure if that has an effect. The `TileLayer.view` instances are added to the scene `Sprite`, which is a child of `Main`. 

The issue that I observed when setting `scaleX` and `scaleY` was the non-square pixels. I'll try to experiment with the properties that Joshua mentioned, and see what happens.

**Comment by cactusanctuary** [comment]
I've tried a couple of things out and got some fun results. First, here is a screenshot with my game at scale 3.3333333333, to fill up my screen's 1280x800 resolution (up from natural resolution of 384x240). You can notice the non-square pixels for the stars, which is what I do *not* want. What I *do* want is the smoothing as can be seen for the text in the menu at the bottom: it's primarily blue, but on some border pixels it's been smoothed:

![screen shot 2016-05-14 at 10 44 45 pm](https://cloud.githubusercontent.com/assets/1223518/15272352/d0ac4350-1a74-11e6-9775-b4dffee2aec7.png)

I tried setting the `TileLayer.useSmoothing = true` to see how that affects the rendering. The result isn't very nice, it's lost all crispness. I guess it's doing the smoothing at the small resolution, and then scales the smoothed result up:

![screen shot 2016-05-15 at 8 05 40 am](https://cloud.githubusercontent.com/assets/1223518/15272358/01020bd4-1a75-11e6-9316-7da9c7a3791b.png)

I also tried routing the `TileLayer` rendered result into a `BitmapData` and using the `scaleX`, `scaleY`, and `smoothing` properties to achieve the smoothing. I hope I understood @jgranick correctly! Code paraphrased here: 
```
public function new() {
  _bitmapData = new BitmapData(ScreenConstants.BASE_WIDTH, ScreenConstants.BASE_HEIGHT);
  var bitmap = new Bitmap(_bitmapData);
  bitmap.smoothing = true;
  bitmap.scaleX = ScreenConstants.SCALE;
  bitmap.scaleY = bitmap.scaleX;
  addChild(bitmap);
}
...
private function update(e:Event):Void {
  _layer.render();
  _bitmapData.draw(_layer.view);
}
```
Screenshot below... So it's rendering from a different spritesheet for whatever reason. Interesting... But besides that, the smoothing is still the same quality as turning on the smoothing on the `TileLayer` itself. 

![screen shot 2016-05-15 at 8 39 27 am](https://cloud.githubusercontent.com/assets/1223518/15272435/a95f90aa-1a78-11e6-841f-c3f9d3d92bdd.png)

So is the post-process scaler shader still the best viable option? The alternative I can think of is, the first time starting up the game, scale the natural resolution images up using nearest neighbor and save those back to the directory, and then use the scaled images with smoothing.

**Comment by jgranick** [comment]
it sounds as if nearest-neighbor is not ideal (due to non-square pixels), as well as bilinear not being ideal (due to fuzzing of edges), so what you want is a custom nearest-neighbor-style implementation?

**Comment by cactusanctuary** [comment]
Yes, something between nearest neighbor and bilinear is what I would like. Shovel Knight did this to achieve a NES resolution in high definition has a natural resolution of 400x240, scaled up 4.5x to 1080p. Lars shows the effect in [some nice screenshots](http://www.fortressofdoors.com/doing-an-hd-remake-the-right-way/#resolvingresolution). 

How could I implement such a scaling effect? Should it indeed be a post-processing effect?

**Comment by larsiusprime** [comment]
I believe Shovel knight is just bilinear. The key is that it's 4.5 : 1 pixel ratio that makes it "crisp enough"

**Comment by cactusanctuary** [comment]
I just tried my game with manually set 2.5 scale, and then 4.5, but it looks just as smooth as 3.333333333. So I don't think it's that. I've asked one of the Shovel Knight programmers how they did it, hopefully he'll respond. 

Otherwise, I'm back to scaling up by NN by the whole number, and then bilinear by the remaining fraction. So in my example, 3x NN, and then 1.1111111111111x bilinear. It sounds like it would be more efficient to do the 3x NN of the assets in one time when opening the game for the first time, instead of doing both scalings each frame. 

Any other ideas?

**Comment by cactusanctuary** [comment]
I'm going to try to do the above very quickly, possibly using yupswing's implementation of post-processing. If I can do it, would @jgranick like me to make a pull request for it in OpenFL somewhere?

**Comment by larsiusprime** [comment]
> Shovel Knight is rendered up to nearest int using NN, then bilinear is applied in post-processing for the remaining scale. 

Very clever!

**Comment by cactusanctuary** [comment]
I tried yupswing's `PostFX` class but it can't compile, neither for Legacy nor Next. 

So I tried creating a `BitmapData` out of my `Scene` instance, adding a `Bitmap` made using the `BitmapData`, and then using `_bitmapData.draw(_scene)` to render it. Unfortunately it seems `TileLayer` does *not* work well with `BitmapData`, because I keep getting funny rendering like. I tried a few different implementations, but no luck:

`_bitmapData.draw(_scene, null, null, null, null, false)` where `smoothing = false` but it still smoothes the result. Same if `smoothing = true`.
![screen shot 2016-05-18 at 7 42 29 pm](https://cloud.githubusercontent.com/assets/1223518/15368963/c445ee54-1d30-11e6-8c30-23092f6437c2.png)

I also tried setting the `smoothing = true` on the `Bitmap` instance itself, but the result is the same as having it `false`. 

I also tried setting `_scene.cacheAsBitmap = true` but no change either. 

My feeling right now is that this would take too long to implement for my need. If anybody else has any ideas that would be quick to try out, I would love to hear them.
